### ISSUE TITLE
SystemStubsExtension not working well with QuarkusTest

### ISSUE THREAD

**Issue by PierreBtz** [webcompere/system-stubs#48]
Hello, I noticed that when working with a `QuarkusTest`, the `SystemStubsExtension` does not work properly. 

## Issues

1. Field injection is not properly working:

```
@QuarkusTest
@ExtendWith(SystemStubsExtension.class)
class SystemStubsQuarkusTest {

    @SystemStub
    EnvironmentVariables environmentVariables;

    @Test
    void environmentVariableInFieldShouldBeSet() {
        environmentVariables.set("FOO", "BAR");
        assertEquals("BAR", System.getenv("FOO"));
    }
}
```

`environmentVariables` does not get injected and the test fails with:

> NullPointer Cannot invoke "uk.org.webcompere.systemstubs.environment.EnvironmentVariables.set(String, String)" because "this.environmentVariables" is null

2. Method parameter injection not properly working either:

```
@QuarkusTest
@ExtendWith(SystemStubsExtension.class)
class SystemStubsQuarkusTest {

    @Test
    void environmentVariableInMethodParameterShouldBeSet(EnvironmentVariables env) {
        env.set("FOO1", "BAR1");
        assertEquals("BAR1", System.getenv("FOO1"));
    }
}
```

The variable seems to be injected (env is not null) BUT the value is not set:

> expected: <BAR1> but was: <null>

## Ideas

I think that for 1. it might be the same root issue as with a `@TempDir`, ie QuarkusTest are creating their own objects, as explained in this comment: https://github.com/quarkusio/quarkus/issues/13261#issuecomment-727902502.

In case 2, the injection is done, but then setting variables is failing. I didn't analyze this further...

## Reproduction

For convenience, I created a sample app to quickly reproduce:
1. clone https://github.com/PierreBtz/quarkus-system-stubs-bug
2. run `./mvnw verify` and observe the failure. The failing tests are in https://github.com/PierreBtz/quarkus-system-stubs-bug/blob/main/src/test/java/eu/pierrebeitz/SystemStubsQuarkusTest.java. For comparison the exact same tests but running as "standard" junit tests are in https://github.com/PierreBtz/quarkus-system-stubs-bug/blob/main/src/test/java/eu/pierrebeitz/SystemStubsTest.java

**Comment by ashleyfrieze** [comment]
@PierreBtz thanks for the report. By the looks of it, QuarkusTest hacks JUnit quite a lot. There are some possible workarounds:

- Try `@Nested` tests - see if the nested test can have both Quarkus AND System Stubs in it
- Don't use the Extension, but instead make a new `EnvironmentVariables` object in the initializer list, and then use the `@BeforeEach` and `@AfterEach` method of the class to perform the setup/teardown on the object - as described here - https://github.com/webcompere/system-stubs#usage-of-individual-system-stubs

**Comment by PierreBtz** [comment]
> Don't use the Extension, but instead make a new EnvironmentVariables object in the initializer list, and then use the @BeforeEach and @AfterEach method of the class to perform the setup/teardown on the object - as described here - [webcompere/system-stubs#usage-of-individual-system-stubs](https://github.com/webcompere/system-stubs#usage-of-individual-system-stubs)

This was indeed my workaround, sorry I should have mentioned that :)

> Try @Nested tests - see if the nested test can have both Quarkus AND System Stubs in it

I quickly checked, but sadly it doesn't work, or at least I could not get it to work. Not that big of a deal, using "the manual way" is good enough for my use case.

**Comment by ashleyfrieze** [comment]
Hey @PierreBtz - if Quarkus is abusing the JUnit5 lifecycle, then there's not much an innocent plugin can do about that...

However, if you'd like to write some changes to the README.md (possibly link to some Quarkus specific advice in a sub .md file) and pitch them as a PR, I'd be happy to try to get that published.

I don't see what options we have for changing the SystemStubs extension here....

**Comment by PierreBtz** [comment]
@ashleyfrieze I'll try to come up with something, it's indeed worth to document.

I just discovered something else, Quarkus also has some funky stuff going on with classloading (see for example https://github.com/quarkusio/quarkus/issues/12511) and this totally messes up with `SystemStubs` if you use it both in a Quarkus and non Quarkus test in the same execution. Basically what happens is that depending on the test (Quarkus/NonQuarkus), the initialization code will execute in two different classloader, which means https://github.com/webcompere/system-stubs/blob/93db7a102bb92c832561e251b15caa747eb1c4a6/system-stubs-core/src/main/java/uk/org/webcompere/systemstubs/environment/EnvironmentVariableMocker.java#L26-L49 will be called twice, which means:

```
Caused by: java.lang.IllegalStateException: Cannot set up environment mocking:
For java.lang.ProcessEnvironment, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered. This may be a result of not having the right mockito-inline installed, or may be down to a Java internals change.
	at uk.org.webcompere.systemstubs.environment.EnvironmentVariableMocker.<clinit>(EnvironmentVariableMocker.java:45)
	... 60 more
Caused by: org.mockito.exceptions.base.MockitoException:
For java.lang.ProcessEnvironment, static mocking is already registered in the current thread

To create a new mock, the existing static mock registration must be deregistered
	at uk.org.webcompere.systemstubs.environment.EnvironmentVariableMocker.<clinit>(EnvironmentVariableMocker.java:29)
	... 60 more
```

Fun stuff :)

I guess the simplest advice if the production code cannot be changed would be:
* do not use the extension
* split your `Quarkus` and `QuarkusTest` into different executions.

But in my case, I had a look to the various calls and with some slight refactoring, I decided to wrap the static calls to `System` into a bean and mock it during tests.

I still a few places where I need to use `System` and therefore `SystemStubs` (because of third parties) but by definition, I don't need QuarkusTest for those so I'm all good :)

I'll come up with some documentation.
